### PR TITLE
fix random attention for pytorch's bigbird/pegasus_bigbird

### PR DESCRIPTION
**Pull request by Bearnardd**
Fixes # (issue)
https://github.com/huggingface/transformers/issues/23055

# What does this PR do?
Add control over usage of random attention of `BigBird` based on current mode (training/eval)

## Who can review?

@sanchit-gandhi @ydshieh

### CODE REVIEW

**Comment by ydshieh** [comment]
Hi @Bearnardd Thank you for the PR.

I have one question: why `def _bigbird_block_rand_mask_with_head` is not modified for this pytorch BigBird file ..?

File: src/transformers/models/big_bird/modeling_big_bird.py [hunk 1]
```diff
@@ -568,7 +568,13 @@ def bigbird_block_sparse_attention(
         if from_seq_len in [1024, 3072, 4096]:  # old plans used in paper
             rand_attn = [
                 self._bigbird_block_rand_mask(
-                    self.max_seqlen, self.max_seqlen, from_block_size, to_block_size, n_rand_blocks, last_idx=1024
+                    self.max_seqlen,
+                    self.max_seqlen,
+                    from_block_size,
+                    to_block_size,
+                    n_rand_blocks,
+                    self.training,
```
Comment (sanchit-gandhi) [inline src/transformers/models/big_bird/modeling_big_bird.py:576]: Looks like we set `deterministic=self.training`. Don't think this is quite right, should be the other way round, e.g.
```suggestion
                    not self.training,
```

File: src/transformers/models/big_bird/modeling_big_bird.py [hunk 2]
```diff
@@ -1077,6 +1084,9 @@ def _bigbird_block_rand_mask(
             raise ValueError("Error the number of blocks needs to be same!")
 
         rand_attn = np.zeros((from_seq_length // from_block_size - 2, num_rand_blocks), dtype=np.int32)
+        # During inference (eval) no randomness
+        if deterministic:
+            return rand_attn
```
Comment (sanchit-gandhi) [inline src/transformers/models/big_bird/modeling_big_bird.py:1089]: The alternative would be to rename `deterministic` to `training` everywhere in this function (this would be fine in PyTorch, but we prefer `deterministic` in Flax!)
```suggestion
        # During inference (eval) no randomness
        if not training:
            return rand_attn
```

File: src/transformers/models/big_bird/modeling_big_bird.py [hunk 3]
```diff
@@ -1160,11 +1170,17 @@ def _bigbird_block_rand_mask_with_head(
...
         # till when to follow plan
         max_plan_idx = plan_from_length.index(from_seq_length)
+
         # Random Attention adjacency list
         rand_attn = [
             np.zeros((num_blocks, np.sum(plan_num_rand_blocks[: max_plan_idx + 1])), dtype=np.int32)
             for i in range(num_heads)
         ]
+        # During inference (eval) no randomness
+        if not self.training:
```
Comment (sanchit-gandhi) [inline src/transformers/models/big_bird/modeling_big_bird.py:1180]: The above comment is the same as what you do here!

File: src/transformers/models/big_bird/modeling_big_bird.py [hunk 4]
```diff
@@ -1054,7 +1060,7 @@ def _get_rand_attn_plan(from_seq_length, from_block_size, num_rand_blocks):
 
     @staticmethod
     def _bigbird_block_rand_mask(
-        from_seq_length, to_seq_length, from_block_size, to_block_size, num_rand_blocks, last_idx=-1
+        from_seq_length, to_seq_length, from_block_size, to_block_size, num_rand_blocks, deterministic, last_idx=-1
```
Comment (sanchit-gandhi) [inline src/transformers/models/big_bird/modeling_big_bird.py:1063]: Wonder if it's better to make this a non-static method? E.g. so that we can do `self.training` from within the method and not have to pass `deterministic` as an argument? Just a suggestion, fine either way for me.

**Review by sanchit-gandhi — COMMENTED** [review]
Very cool @Bearnardd - thanks for jumping on fixing the PyTorch model so quickly! Just a few small suggestions / fixes from me!

File: src/transformers/models/big_bird/modeling_big_bird.py (hunk 4 above)
Comment (Bearnardd) [reply src/transformers/models/big_bird/modeling_big_bird.py:1063]: @sanchit-gandhi yeah I was thinking about it and that is why I asked about `@staticmethod` for this method :)

File: src/transformers/models/big_bird/modeling_big_bird.py (hunk 2 above)
Comment (Bearnardd) [reply src/transformers/models/big_bird/modeling_big_bird.py:1089]: Oh yeah, doing stuff while being on the vacation can be a little bit chaotic :)

**Comment by Bearnardd** [comment]
Hi @sanchit-gandhi! I have removed the static method as I think it is the best approach.

**Comment by Bearnardd** [comment]
> Hi @Bearnardd Thank you for the PR.
> 
> I have one question: why `def _bigbird_block_rand_mask_with_head` is not modified for this pytorch BigBird file ..?

Thanks for the comment! To be honest I am not sure If I understand you correctly, since from what I can see this function is updated. Could you elaborate what exactly is missing?

**Comment by ydshieh** [comment]
> > Hi @Bearnardd Thank you for the PR.
> > I have one question: why `def _bigbird_block_rand_mask_with_head` is not modified for this pytorch BigBird file ..?
> 
> Thanks for the comment! To be honest I am not sure If I understand you correctly, since from what I can see this function is updated. Could you elaborate what exactly is missing?

Sorry, my bad. You are right :-)

File: tests/models/big_bird/test_modeling_big_bird.py
Comment (ydshieh) [inline tests/models/big_bird/test_modeling_big_bird.py:1]: @Bearnardd Do you know why, before this PR, the test like `test_inference_block_sparse_pretraining` could get deterministic outputs? It's somehow strange to me that it is the case as the goal of this PR is to fix 

> random attention is used no matter whether we are in training/eval mode

Comment (Bearnardd) [reply tests/models/big_bird/test_modeling_big_bird.py:1]: Yeah, sure. It has deterministic output because the random seed for random numpy operations is hardcoded which results in the same random attention indices every time you run tests.

Comment (ydshieh) [reply tests/models/big_bird/test_modeling_big_bird.py:1]: Thanks for the explanation!

**Review by sanchit-gandhi — APPROVED** [review]
Awesome - thanks for fixing the big bird models @Bearnardd!

**Comment by Bearnardd** [comment]
cc @sgugger

File: src/transformers/models/big_bird/modeling_big_bird.py [hunk 5]
```diff
@@ -568,7 +568,12 @@ def bigbird_block_sparse_attention(
         if from_seq_len in [1024, 3072, 4096]:  # old plans used in paper
             rand_attn = [
                 self._bigbird_block_rand_mask(
-                    self.max_seqlen, self.max_seqlen, from_block_size, to_block_size, n_rand_blocks, last_idx=1024
+                    self.max_seqlen,
+                    self.max_seqlen,
+                    from_block_size,
+                    to_block_size,
+                    n_rand_blocks,
+                    last_idx=1024,
```
Comment (sgugger) [inline src/transformers/models/big_bird/modeling_big_bird.py:576]: There is no need for the restyle anymore here, might be some leftover from when there was `deterministic` passed?

File: src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py [hunk 6]
```diff
@@ -389,7 +389,12 @@ def bigbird_block_sparse_attention(
         if from_seq_len in [1024, 3072, 4096]:  # old plans used in paper
             rand_attn = [
                 self._bigbird_block_rand_mask(
-                    self.max_seqlen, self.max_seqlen, from_block_size, to_block_size, n_rand_blocks, last_idx=1024
+                    self.max_seqlen,
+                    self.max_seqlen,
+                    from_block_size,
+                    to_block_size,
+                    n_rand_blocks,
+                    last_idx=1024,
```
Comment (sgugger) [inline src/transformers/models/bigbird_pegasus/modeling_bigbird_pegasus.py:397]: Same here

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! Just two small nits and it should be good to merge.

**Comment by Bearnardd** [comment]
I have pushed the changes @sgugger :)

**Review by sgugger — APPROVED** [review]
